lib/src/nodes/config_node_object.cc

```diff
@@ -0,0 +1,320 @@
+#include <internal/nodes/config_node_object.hpp>
+#include <internal/nodes/config_node_field.hpp>
+#include <internal/nodes/config_node_single_token.hpp>
+#include <internal/tokens.hpp>
+#include <internal/path_parser.hpp>
+#include <internal/nodes/config_node_include.hpp>
+#include <boost/range/iterator_range_core.hpp>
+
+using namespace std;
+
+namespace hocon {
+
+    config_node_object::config_node_object(shared_node_list children) :
+            config_node_complex_value(move(children)) { }
+
+    shared_ptr<config_node_complex_value> config_node_object::new_node(shared_node_list nodes) {
+        return make_shared<config_node_object>(move(nodes));
+    }
+
+    bool config_node_object::has_value(path desired_path) const {
+        for (auto&& node : children()) {
+            if (auto field = dynamic_pointer_cast<config_node_field>(node)) {
+                path key = field->path()->get_path();
+                if (key == desired_path || key.starts_with(desired_path)) {
+                    return true;
+                } else if (desired_path.starts_with(key)) {
+                    if (auto object = dynamic_pointer_cast<config_node_object>(field->get_value())) {
+                        path remaning_path = desired_path.sub_path(key.length());
+                        if (object->has_value(remaning_path)) {
+                            return true;
+                        }
+                    }
+                }
+            }
+        }
+        return false;
+    }
+
+    // TODO: Just gonna throw this out here, this method is disgustingly long and hard to read...
+    // (line for line port from Java)
+    shared_ptr<config_node_object> config_node_object::change_value_on_path(path desired_path, shared_node_value value,
```

Comment:
Candidate for refactoring if we have to make significant changes to it (bugs, new features). Probably not worth it until then, because we'd also want to change the other projects.